### PR TITLE
[oneDNN EP] Improves threading performance and adds support ORT and Eigen threadpools

### PR DESCRIPTION
**Pull request by eralmual**
* Added oneDNN support to use an unified threadpool with ORT, users can select this with the --dnnl_thread ort build flag.

* Fixed oversubscription happening with ORT and OpenMP oneDNN threads.

* Added OrtDnnlProviderOptions struct to support provider creation with new features and customization.

* Added option for users to customize the numbers of OpenMP and Eigen threads used by oneDNN.

* perf_test and MNIST training now supports OpenMP, Eigen and ORT threadpools.

* Java and Python support for new threadpool options.

**Description**: Previously OneDNN ep suffered from thread oversubscription due to ORT creating it's own thread pool and OneDNN creating another one, this update aims to fix this issue by using one unified threadpool for both the default provider and oneDNN ep by default. Also enables support for Eigen threadpool and prevents the oversubscription issue when using OpenMP.
The OrtDnnlProviderOptions struct is added, this allows for deeper customization of the oneDNN ep with features that will be coming in the next few moths.

**Motivation and Context**
- Why is this change required? What problem does it solve?
Performance related fix, prevents thread oversubscription, improves performance on CPU and allows the user to customize thread distribution according to their workload.

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by eralmual** [comment]
@jywu-msft maintainer please run `/azp run Linux DNNL CI Pipeline` since this is the primary pipeline for testing the OneDNN execution provider.

**Comment by georgen117** [comment]
@jywu-msft could you please run the CI `/azp run Linux DNNL CI Pipline` for this PR?

**Comment by eralmual** [comment]
@jywu-msft could you try running `/azp run Linux DNNL CI Pipeline`  again please?

**Comment by eralmual** [comment]
Sorry @jywu-msft could you give it another try?

**Comment by eralmual** [comment]
@jywu-msft just to remind you about this

**Comment by eralmual** [comment]
@jywu-msft could you try again please

**Comment by georgen117** [comment]
@jywu-msft would you mind running the `/azp run Linux DNNL CI Pipeline` again?

**Comment by georgen117** [comment]
> No pipelines are associated with this pull request.

@jywu-msft Do you know why we got this message?

**Comment by jywu-msft** [comment]
> > No pipelines are associated with this pull request.
> 
> @jywu-msft Do you know why we got this message?

sorry, i did a copy and paste and it seems the source had a typo.

**Comment by eralmual** [comment]
@jywu-msft could you try running `/azp run Linux DNNL CI Pipeline` again please?

**Comment by eralmual** [comment]
@jywu-msft can we try again please? `/azp run Linux DNNL CI Pipeline`

**Comment by georgen117** [comment]
@jywu-msft can you run  `/azp run Linux DNNL CI Pipeline` again please?

**Comment by eralmual** [comment]
@jywu-msft could you run `/azp run Linux DNNL CI Pipeline` again please?

File: cmake/external/dnnl.cmake [hunk 1]
```diff
@@ -15,6 +15,13 @@ else()
   endif()  
 endif()
 
+if(onnxruntime_USE_DNNL AND ((onnxruntime_DNNL_THREAD STREQUAL "eigen") OR (onnxruntime_DNNL_THREAD STREQUAL "ort")))
+  message(STATUS "dnnl is set to use a custom threadpool thread")
+  set(DNNL_THREAD_CMAKE_ARGS "-DDNNL_CPU_RUNTIME=THREADPOOL")
+else()
+  message(STATUS "dnnl is set to use openmp")
+endif()
+
```
Comment (sorny92) [inline cmake/external/dnnl.cmake:24]: Does it make sense to add the user to build onednn with either OMP or TBB in case they don't use `threadpool`?

Comment (georgen117) [reply cmake/external/dnnl.cmake:24]: @jywu-msft could you run /azp run Linux DNNL CI Pipeline again please?

**Comment by georgen117** [comment]
@jywu-msft could you run /azp run Linux DNNL CI Pipeline again please?

**Comment by eralmual** [comment]
@jywu-msft could you run `/azp run Linux DNNL CI Pipeline` please?

**Comment by georgen117** [comment]
@jywu-msft could you please kick off the other CI builds for this PR?

**Comment by eralmual** [comment]
@jywu-msft just a reminder to please run the other CI builds

**Comment by jywu-msft** [comment]
@pranavsharma can you help review these changes? there's some changes to files in core/session, core/platform etc.

File: onnxruntime/core/session/inference_session.cc [hunk 2]
```diff
@@ -517,7 +517,13 @@ common::Status InferenceSession::RegisterExecutionProvider(const std::shared_ptr
       ORT_RETURN_IF_ERROR(p_exec_provider->SetComputeStream(trt_ep->GetComputeStream()));
     }
   }
-
+#ifndef DNNL_OPENMP
+  // This is safetly called only when we are not using OpenMP, preprocessor directive will probably be
+  // removed in the future
+  if (provider_type == onnxruntime::kDnnlExecutionProvider) {
+    ORT_RETURN_IF_ERROR(p_exec_provider->SetComputeStream(static_cast<void*>(GetIntraOpThreadPoolToUse())));
```
Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.cc:524]: Not sure I'm following this. I suppose you're reusing the void* of the compute stream for the ORT thread pool. This is super confusing and compromises readability.

File: include/onnxruntime/core/platform/threadpool.h [hunk 3]
```diff
@@ -378,6 +375,9 @@ class ThreadPool {
   // thread in the pool. Returns -1 otherwise.
   int CurrentThreadId() const;
 
+ private:
```
[the remark this answers is not in the document]
Comment (eralmual) [reply include/onnxruntime/core/platform/threadpool.h:378]: Yes it is, we need access to those methods in order to interface with oneDNN

**Comment by eralmual** [comment]
@jywu-msft
 @pranavsharma just a reminder about this

File: onnxruntime/core/session/inference_session.cc (hunk 2 above)
Comment (pranavsharma) [reply onnxruntime/core/session/inference_session.cc:524]: We need to think this through. I don't think the changes in this PR to achieve this in the current form are acceptable.

Comment (georgen117) [reply onnxruntime/core/session/inference_session.cc:524]: @pranavsharma we are open to suggestions as well as testing out alternate proposals.  Do you have any suggestions?

**Pull request closed without merging**